File: src/middleware/errorHandler.ts
import type { FastifyError, FastifyReply, FastifyRequest } from "fastify";
import { logger } from "../logger.js";

export function errorHandler(
  error: FastifyError & { name?: string },
  _req: FastifyRequest,
  reply: FastifyReply,
) {
  logger.error({ err: error.message, code: error.statusCode, name: error.name }, "Request error");

  if (error.validation) {
    return reply.status(400).send({ error: "Validation failed", details: error.validation });
  }

  if (error.name === "YtDlpTimeoutError") {
    return reply.status(504).send({ error: "Upstream timed out" });
  }

  if (error.name === "YtDlpError") {
    return reply.status(502).send({ error: "Upstream service error" });
  }

  const status = error.statusCode ?? 500;
  const message = status === 500 ? "Internal server error" : error.message;
  return reply.status(status).send({ error: message });
}

File: server/src/middleware/error_handler.py
from __future__ import annotations

from fastapi import Request, status
from fastapi.responses import JSONResponse

from src.logger import logger
from src.services.ytdlp import YtDlpError, YtDlpTimeoutError


async def ytdlp_timeout_handler(_request: Request, _exc: YtDlpTimeoutError) -> JSONResponse:
    logger.error("yt-dlp timeout")
    return JSONResponse(
        status_code=status.HTTP_504_GATEWAY_TIMEOUT,
        content={"error": "Upstream timed out"},
    )


async def ytdlp_error_handler(_request: Request, exc: YtDlpError) -> JSONResponse:
    logger.error("yt-dlp error: %s", str(exc))
    logger.error("yt-dlp stderr: %s", exc.stderr)
    return JSONResponse(
        status_code=status.HTTP_502_BAD_GATEWAY,
        content={"error": f"Upstream service error: {exc}"},
    )


async def generic_error_handler(_request: Request, exc: Exception) -> JSONResponse:
    logger.exception("Unhandled error: %s", exc)
    return JSONResponse(
        status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
        content={"error": "Internal server error"},
    )
